mirgecom/profiling.py

```diff
@@ -138,48 +93,115 @@ def __init__(self, queue, allocator=None) -> None:
+
+        name = knl.function_name
+
+        args_tuple = tuple(
+            (arg.size)
+            for arg in actual_args if isinstance(arg, cl.array.Array))
+
+        try:
+            ignore = self.kernel_stats[knl][args_tuple]  # noqa
+        except KeyError:
+            nbytes = 0
+            nops = 0
+            for arg in actual_args:
+                if isinstance(arg, cl.array.Array):
+                    nbytes += arg.size * arg.dtype.itemsize
+                    nops += arg.size
+            res = SingleCallKernelProfile(time=0, flops=nops, bytes_accessed=nbytes,
+                                footprint_bytes=nbytes)
+            self.kernel_stats.setdefault(knl, {})[args_tuple] = res
+
+        if self.logmgr and f"{name}_time" not in self.logmgr.quantity_data:
+            self.logmgr.add_quantity(KernelProfile(self, name))
+
+        self.profile_events.append(ProfileEvent(evt, knl, args_tuple))
+
+        return evt
 
     def _finish_profile_events(self) -> None:
         # First, wait for completion of all events
         if self.profile_events:
             cl.wait_for_events([pevt.cl_event for pevt in self.profile_events])
 
-        global nonloopy_profile_events
-        if nonloopy_profile_events:
-            cl.wait_for_events([npe.cl_event for npe in nonloopy_profile_events])
-
         # Then, collect all events and store them
         for t in self.profile_events:
             program = t.program
             r = self._get_kernel_stats(program, t.args_tuple)
             time = t.cl_event.profile.end - t.cl_event.profile.start
 
-            new = ProfileResult(time, r.flops, r.bytes_accessed, r.footprint_bytes)
-
-            self.profile_results.setdefault(program, []).append(new)
+            new = SingleCallKernelProfile(time, r.flops, r.bytes_accessed,
+                                          r.footprint_bytes)
 
-        for t in nonloopy_profile_events:
-            program = t.knl
-            time = t.cl_event.profile.end - t.cl_event.profile.start
-            new = ProfileResult(time, t.nops, t.nbytes, t.nbytes)
             self.profile_results.setdefault(program, []).append(new)
 
         self.profile_events = []
-        nonloopy_profile_events = []
 
-    def tabulate_profiling_data(self) -> pytools.Table:
+    def get_and_reset_profiling_data_for_kernel(self, kernel_name: str,
+                                 wait_for_events=True) -> MultiCallKernelProfile:
+        """Return and reset profiling data for kernel `kernel_name`."""
+        if wait_for_events:
+            self._finish_profile_events()
+
+        def _gather_data(results_dict: dict):
+            results = [key for key in results_dict if hasattr(key, "name")
+                       and key.name == kernel_name]
+            results.extend(
+                [key for key in results_dict if hasattr(key, "function_name")
+                 and key.function_name == kernel_name])
```

Comment:
These traversals could get quite expensive. Why not insert them into the dictionary under their name in the first place if that's all we care about in the end?

Reply:
I've simplified the code a bit; the reason not to index them by name is that sometimes we want to filter by name+kwargs

Reply:
Not sure I understand. Could you give an example?

Reply:
In `tabulate_profiling_data()`, we separate kernels based on the actual kernel object, as well as the `args_tuple` (=`kwargs`) that the kernel was called with in order to differentiate between different kernels that might have the same name. In the logging infrastructure, this is impractical, and we only look at the kernel name.

I've simplified and improved the code for `get_and_reset_profiling_data_for_kernel` a bit more.